### PR TITLE
Origin/feature bonus move

### PR DESCRIPTION
**Pull request by NamesAreHard6600**
# Overview
Adds a new scan for the bonus movement after firing a weapon like with Chen Rong. The actual address has a few caveats with how the game handles this mechanic i.e. you have to add the bonus move to a mech after it fires its weapon or it will not get its action, only the bonus move (which could be what you wanted in some cases). 

# What This Scan Provides
This scan does 2 things:
1. It helps fix a few bugs, in particular, reactivating mechs with Chen or Archie as a pilot. When reactivating these mechs, the bonus move overrides the reactivation since the mech is still considered "active". This means that the mech does not get its bonus move. This scan can fix that, by first setting their bonus move to zero to cancel the pilot ability, and then resetting them to active, giving the turn back. After firing the weapon again, Chen would reapply their buff and the mech would get its bonus move. 
2. It also, (as many scans do) provides design space for modders to make new things. This scan would allow new pilots with bonus move abilities, weapons that let you move after firing them, passives that provide it to your whole team, and more. 

# Issue
The scan is fully functional and consistent in its working. The issue is that it requires the player to bring Chen Rong in as one of their pilots. While not seeming like a huge deal, this causes a few problems:
1. Other scans start also requiring pilots; Suggested on this topic was using console commands to grab the pilots instead
2. Someone does not have Chen Rong unlocked on their profile and now must additionally use console commands to do so, if not reverting to a console command solution
3. As well as just a general hassle when scanning memedit that wasn't there before. 

## My Opinion
Personally, with the infrequent scans that are required (we don't really see Subset updating the game anymore) I think it's not an issue, people can figure it out. In addition to that, after an update is released by subset, a modloader update is usually followed, meaning the addresses can be passed through the modloader and anyone struggling with the process can wait for that, or request the address file from someone who has it. Overall, the process isn't *that* complicated. It can just take a bit of extra time. 

Once we get some other opinions, we can decide together whether we should add the scan or not.

### CODE REVIEW

File: scanner/utils.lua [hunk 1]
```diff
@@ -19,7 +19,7 @@ end
 
 function utils.boardExists()
 	if Board == nil or GetCurrentMission() == nil then
-		return false, "Enter a mission"
+		return false, "Enter a Mission with Chen as a Pilot"
```
Comment (Lemonymous) [inline scanner/utils.lua:22]: I don't really agree with changing the boardExists function that so many scans use, to request Chen as a pilot. I can understand the idea; you're going to need Chen at some point anyways. Still, each instruction is made short and accurate, which is why there are two different instructions for "mission" vs "real mission".

File: scanner/utils.lua [hunk 2]
```diff
@@ -29,9 +29,9 @@ end
 
 function utils.missionBoardExists()
 	if Board == nil or GetCurrentMission() == nil then
-		return false, "Enter a mission"
+		return false, "Enter a Mission with Chen as a Pilot"
```
Comment (Lemonymous) [inline scanner/utils.lua:32]: Same as for utils.boardExist. I believe this is out of scope of the function.

File: scanner/utils.lua [hunk 3]
```diff
@@ -29,9 +29,9 @@ end
...
 	elseif Board:IsMissionBoard() == nil then
-		return false, "Enter a real mission"
+		return false, "Enter a Real Mission with Chen as a Pilot"
```
Comment (Lemonymous) [inline scanner/utils.lua:34]: Same as above

File: scanner/pawn.lua [hunk 4]
```diff
@@ -73,6 +74,63 @@ scans.baseMaxHealth = inheritClass(Scan, {
 	end
 })
 
+scans.bonusMove = inheritClass(Scan, {
+    id = "BonusMove",
```
Comment (Lemonymous) [inline scanner/pawn.lua:78]: The whitespace (spaces/tabs) used in this function is quite different from the rest of memedit's code.

memedit uses tab for indentation, and no tab at empty lines.

**Review by Lemonymous — COMMENTED** [review]
Looks very good!
I have some feedback in various places.
As for a cleanup function for the scan. I could probably take a stab at it later if I find time.

File: scanner/utils.lua (hunk 1 above)
Comment (NamesAreHard6600) [reply scanner/utils.lua:22]: That's fair, and that's the reason I put it in a separate commit. Should be easy to revert.

File: scanner/pawn.lua (hunk 4 above)
Comment (NamesAreHard6600) [reply scanner/pawn.lua:78]: Yup, that's my bad, a copy paste error on my part.

**Comment by NamesAreHard6600** [comment]
Okay I believe the issues pointed out have been fixed. I will continue thinking of a (clean) solution for the cleanup as well.

**Comment by WolframParadoxica** [comment]
Just a minor thing:
`This means that the mech does not get its bonus _**action**_.`
